File: frontend/src/app/core/services/auth.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable, computed, inject, signal } from '@angular/core';
import { Observable, catchError, map, of, tap } from 'rxjs';
import {
  ApiResponse,
  AuthResponse,
  CompleteProfileRequest,
  GenericAcknowledgementResponse,
  LoginRequest,
  PasswordChangeRequiredResponse,
  ProfileCompletionRequiredResponse,
  RegisterRequest,
  TwoFactorRequiredResponse,
} from '../models/auth.models';
import { TenantMembershipSummary } from '../models/organization.models';

const SESSION_STORAGE_KEY = 'ten21_auth_session';

/**
 * US-11: reactive session state + the three auth HTTP calls. The refresh token itself
 * never touches this service or localStorage — it lives only in the ten21_refresh_token
 * HTTP-only cookie the browser attaches automatically (withCredentials: true) per
 * SECURITY.md §2. What IS kept here/in localStorage is the short-lived (15 min) access
 * token and its accompanying claims, so a page reload doesn't force a fresh login while
 * the refresh cookie is still valid.
 */
@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly http = inject(HttpClient);

  private readonly _session = signal<AuthResponse | null>(this.readStoredSession());
  readonly session = this._session.asReadonly();

  /** US-15: holds the interim (profile_incomplete) token between /api/auth/google
   * returning ProfileCompletionRequiredResponse and the CompleteProfile page submitting
   * it -- deliberately memory-only (not localStorage), since it's short-lived and
   * single-purpose, never a real session. */
  private readonly _interimToken = signal<string | null>(null);
  readonly interimToken = this._interimToken.asReadonly();

  /** US-17: holds the pending 2FA challenge between login() returning
   * TwoFactorRequiredResponse and the VerifyTwoFactor page submitting a code -- same
   * memory-only, single-purpose reasoning as _interimToken above. Kept as its own signal
   * (not folded into _interimToken) purely to keep the two flows' state distinct. */
  private readonly _twoFactorChallenge = signal<TwoFactorRequiredResponse | null>(null);
  readonly twoFactorChallenge = this._twoFactorChallenge.asReadonly();

  /** US-24: holds the pending MustChangePassword challenge between login() returning
   * PasswordChangeRequiredResponse and the ChangeTempPassword page submitting a new
   * password -- same memory-only, single-purpose reasoning as _twoFactorChallenge. */
  private readonly _passwordChangeChallenge = signal<PasswordChangeRequiredResponse | null>(null);
  readonly passwordChangeChallenge = this._passwordChangeChallenge.asReadonly();

  readonly isAuthenticated = computed(() => {
    const session = this._session();
    return session !== null && new Date(session.expiresAtUtc).getTime() > Date.now();
  });

  readonly role = computed(() => this._session()?.role ?? null);
  readonly tenantId = computed(() => this._session()?.tenantId ?? null);
  readonly organizationId = computed(() => this._session()?.organizationId ?? null);
  readonly accessToken = computed(() => this._session()?.accessToken ?? null);

  /** US-17/US-24: returns a full AuthResponse (session set immediately, as before), a
   * TwoFactorRequiredResponse (mandatory-role account -- the challenge is stashed for
   * VerifyTwoFactor), or a PasswordChangeRequiredResponse (account provisioned with a
   * temporary password -- the challenge is stashed for ChangeTempPassword). Checked in
   * that order to match AuthController's own precedence (MustChangePassword is checked
   * before 2FA). */
  login(request: LoginRequest): Observable<AuthResponse | TwoFactorRequiredResponse | PasswordChangeRequiredResponse> {
    return this.http
      .post<ApiResponse<AuthResponse | TwoFactorRequiredResponse | PasswordChangeRequiredResponse>>(
        '/api/auth/login',
        request,
        { withCredentials: true },
      )
      .pipe(
        map((response) => response.data!),
        tap((result) => {
          if ('requiresPasswordChange' in result) {
            this._passwordChangeChallenge.set(result);
          } else if ('requiresTwoFactor' in result) {
            this._twoFactorChallenge.set(result);
          } else {
            this.setSession(result);
          }
        }),
      );
  }

  /** US-24: submits a new password against the stashed MustChangePassword challenge --
   * same "use the challenge token directly as the Authorization header" pattern as
   * verifyTwoFactor()/completeProfile(), since there's no session yet. */
  changeTempPassword(newPassword: string): Observable<AuthResponse> {
    const challenge = this._passwordChangeChallenge();
    if (!challenge) {
      throw new Error('changeTempPassword() called with no pending challenge -- start over at /login.');
    }

    return this.http
      .post<ApiResponse<AuthResponse>>(
        '/api/auth/change-temp-password',
        { newPassword },
        { withCredentials: true, headers: new HttpHeaders({ Authorization: `Bearer ${challenge.challengeToken}` }) },
      )
      .pipe(
        map((response) => response.data!),
        tap((session) => {
          this._passwordChangeChallenge.set(null);
          this.setSession(session);
        }),
      );
  }

  /** US-17: submits a code against the stashed challenge -- the challenge token is used
   * directly as the Authorization header, same pattern as completeProfile(), since there's
   * no session yet to drive the normal interceptor path. */
  verifyTwoFactor(code: string): Observable<AuthResponse> {
    const challenge = this._twoFactorChallenge();
    if (!challenge) {
      throw new Error('verifyTwoFactor() called with no pending challenge -- start over at /login.');
    }

    return this.http
      .post<ApiResponse<AuthResponse>>(
        '/api/auth/login/verify-2fa',
        { code },
        { withCredentials: true, headers: new HttpHeaders({ Authorization: `Bearer ${challenge.challengeToken}` }) },
      )
      .pipe(
        map((response) => response.data!),
        tap((session) => {
          this._twoFactorChallenge.set(null);
          this.setSession(session);
        }),
      );
  }

  /** US-14: workspace registration. Instant provisioning -- succeeds with the same
   * full-session shape as login(), no separate confirmation step required to start using
   * the product. */
  register(request: RegisterRequest): Observable<AuthResponse> {
    return this.http
      .post<ApiResponse<AuthResponse>>('/api/auth/register', request, { withCredentials: true })
      .pipe(
        map((response) => response.data!),
        tap((session) => this.setSession(session)),
      );
  }

  /** US-15: Google Sign-In. Returns either a full AuthResponse (existing account with a
   * workspace -- session is set immediately, same as login()) or a
   * ProfileCompletionRequiredResponse (first-time signup with no workspace yet -- no
   * session is set; the interim token is stashed for CompleteProfile to use next). */
  loginWithGoogle(idToken: string): Observable<AuthResponse | ProfileCompletionRequiredResponse> {
    return this.http
      .post<ApiResponse<AuthResponse | ProfileCompletionRequiredResponse>>(
        '/api/auth/google',
        { idToken },
        { withCredentials: true },
      )
      .pipe(
        map((response) => response.data!),
        tap((result) => {
          if ('requiresProfileCompletion' in result) {
            this._interimToken.set(result.interimToken);
          } else {
            this.setSession(result);
          }
        }),
      );
  }

  /** US-15: the second half of a first-time Google signup. Uses the stashed interim token
   * directly as the Authorization header rather than going through the normal
   * session-based interceptor path -- there is no session yet at this point. */
  completeProfile(request: CompleteProfileRequest): Observable<AuthResponse> {
    const interimToken = this._interimToken();
    if (!interimToken) {
      throw new Error('completeProfile() called with no interim token -- start over at /login.');
    }

    return this.http
      .post<ApiResponse<AuthResponse>>('/api/auth/complete-profile', request, {
        withCredentials: true,
        headers: new HttpHeaders({ Authorization: `Bearer ${interimToken}` }),
      })
      .pipe(
        map((response) => response.data!),
        tap((session) => {
          this._interimToken.set(null);
          this.setSession(session);
        }),
      );
  }

  /** US-16. */
  resendActivation(email: string): Observable<GenericAcknowledgementResponse> {
    return this.http
      .post<ApiResponse<GenericAcknowledgementResponse>>('/api/auth/resend-activation', { email })
      .pipe(map((response) => response.data!));
  }

  /** US-16. */
  activate(userId: string, token: string): Observable<GenericAcknowledgementResponse> {
    return this.http
      .post<ApiResponse<GenericAcknowledgementResponse>>('/api/auth/activate', { userId, token })
      .pipe(map((response) => response.data!));
  }

  /** US-16. */
  forgotPassword(email: string): Observable<GenericAcknowledgementResponse> {
    return this.http
      .post<ApiResponse<GenericAcknowledgementResponse>>('/api/auth/forgot-password', { email })
      .pipe(map((response) => response.data!));
  }

  /** US-16. */
  resetPassword(userId: string, token: string, newPassword: string): Observable<GenericAcknowledgementResponse> {
    return this.http
      .post<ApiResponse<GenericAcknowledgementResponse>>('/api/auth/reset-password', {
        userId,
        token,
        newPassword,
      })
      .pipe(map((response) => response.data!));
  }

  /** US-28: every workspace the caller holds a TenantMembership in, across all tenants --
   * powers the top-nav workspace switcher dropdown. */
  listWorkspaces(): Observable<TenantMembershipSummary[]> {
    return this.http
      .get<ApiResponse<TenantMembershipSummary[]>>('/api/organization/tenants')
      .pipe(map((response) => response.data!));
  }

  /** US-28: switches the active workspace. Reuses setSession() (same as login/refresh), so
   * every signal derived from the session (tenantId, role, organizationId, isAuthenticated)
   * updates reactively for free -- no separate "workspace" signal needed on top of what
   * already exists. Refreshing already-loaded ROUTE DATA (e.g. a property list fetched once
   * in ngOnInit) is a separate concern the caller (AppHeader) handles by re-navigating to
   * the current route, since Angular signals don't retroactively re-trigger a one-time HTTP
   * call just because a value they never subscribed to changed. */
  switchWorkspace(tenantId: string): Observable<AuthResponse> {
    return this.http
      .post<ApiResponse<AuthResponse>>('/api/organization/switch-context', { tenantId }, { withCredentials: true })
      .pipe(
        map((response) => response.data!),
        tap((session) => this.setSession(session)),
      );
  }

  /** Used by the auth interceptor to silently mint a new access token from the refresh cookie. */
  refresh(): Observable<AuthResponse> {
    return this.http
      .post<ApiResponse<AuthResponse>>('/api/auth/refresh-token', {}, { withCredentials: true })
      .pipe(
        map((response) => response.data!),
        tap((session) => this.setSession(session)),
      );
  }

  logout(): Observable<void> {
    return this.http.post<void>('/api/auth/revoke-token', {}, { withCredentials: true }).pipe(
      map(() => void 0),
      tap(() => this.clearSession()),
      catchError(() => {
        // Best-effort server-side revocation — clear local state regardless so the UI
        // never gets stuck "logged in" because the network call happened to fail.
        this.clearSession();
        return of(void 0);
      }),
    );
  }

  setSession(session: AuthResponse): void {
    this._session.set(session);
    localStorage.setItem(SESSION_STORAGE_KEY, JSON.stringify(session));
  }

  clearSession(): void {
    this._session.set(null);
    localStorage.removeItem(SESSION_STORAGE_KEY);
  }

  private readStoredSession(): AuthResponse | null {
    try {
      const raw = localStorage.getItem(SESSION_STORAGE_KEY);
      return raw ? (JSON.parse(raw) as AuthResponse) : null;
    } catch {
      return null;
    }
  }
}

File: src/Ten21.Business/Auth/AuthService.cs
using System.Security.Cryptography;
using System.Text;
using Microsoft.AspNetCore.Identity;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Ten21.Application.Abstractions;
using Ten21.Domain.Common;
using Ten21.Domain.Entities;
using Ten21.Domain.Exceptions;
using Ten21.Infrastructure.Identity;
using Ten21.Infrastructure.Persistence;

namespace Ten21.Business.Auth;

/// <summary>
/// US-02: Identity & Refresh Token Pipeline.
///
/// Business-layer refactor: all business logic AND all data access now live here --
/// AuthController has no Ten21DbContext dependency at all. Everything that genuinely
/// requires the live HTTP request (ClaimsPrincipal claim reads, the refresh-token cookie
/// itself, HttpContext.Connection.RemoteIpAddress) stays in the controller and is passed in
/// here as plain values (Guid userId, string? clientIp, a raw token string, etc).
///
/// Deliberately uses UserManager directly rather than SignInManager -- SignInManager is
/// built around cookie-based interactive sign-in flows; this is a stateless JWT API, and
/// UserManager.CheckPasswordAsync + manual AccessFailedAsync/ResetAccessFailedCountAsync
/// calls give the same brute-force lockout behavior (SECURITY.docx §5) without pulling in
/// cookie-auth-scheme machinery this API doesn't otherwise use.
/// </summary>
public class AuthService
{
    private const string GenericLoginFailureMessage = "Invalid email or password.";
    private const string GoogleLoginProvider = "Google";

    private readonly UserManager<ApplicationUser> _userManager;
    private readonly RoleManager<ApplicationRole> _roleManager;
    private readonly Ten21DbContext _dbContext;
    private readonly IJwtTokenService _jwtTokenService;
    private readonly IRefreshTokenService _refreshTokenService;
    private readonly ITenantContext _tenantContext;
    private readonly ITurnstileVerificationService _turnstileVerificationService;
    private readonly IGoogleIdTokenVerifier _googleIdTokenVerifier;
    private readonly IEmailSender _emailSender;
    private readonly string _frontendBaseUrl;

    public AuthService(
        UserManager<ApplicationUser> userManager,
        RoleManager<ApplicationRole> roleManager,
        Ten21DbContext dbContext,
        IJwtTokenService jwtTokenService,
        IRefreshTokenService refreshTokenService,
        ITenantContext tenantContext,
        ITurnstileVerificationService turnstileVerificationService,
        IGoogleIdTokenVerifier googleIdTokenVerifier,
        IEmailSender emailSender,
        IConfiguration configuration)
    {
        _userManager = userManager;
        _roleManager = roleManager;
        _dbContext = dbContext;
        _jwtTokenService = jwtTokenService;
        _refreshTokenService = refreshTokenService;
        _tenantContext = tenantContext;
        _turnstileVerificationService = turnstileVerificationService;
        _googleIdTokenVerifier = googleIdTokenVerifier;
        _emailSender = emailSender;
        _frontendBaseUrl = configuration["Frontend:BaseUrl"]?.TrimEnd('/') ?? "http://localhost:4200";
    }

    /// <summary>
    /// US-14: Workspace Registration & Onboarding. Creates the ApplicationUser, a brand-new
    /// Tenant (the "workspace"), and a root TenantMembership with BOTH PropertyManager and
    /// PropertyOwner claims -- a self-service landlord is both the operator and the deed
    /// owner of their own portfolio.
    /// </summary>
    public async Task<AuthTokenResult> RegisterAsync(
        RegisterRequest request, string? clientIp, CancellationToken cancellationToken)
    {
        // US-18: bot defense gates first, before any other validation runs.
        if (!await _turnstileVerificationService.VerifyAsync(request.TurnstileToken, clientIp, cancellationToken))
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                [nameof(request.TurnstileToken)] = ["Bot verification failed. Please try again."],
            });
        }

        if (!request.AgreedToTerms)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                [nameof(request.AgreedToTerms)] = ["You must agree to the Terms of Service to register."],
            });
        }

        if (await _userManager.FindByEmailAsync(request.Email) is not null)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                [nameof(request.Email)] = ["An account with this email already exists."],
            });
        }

        var user = new ApplicationUser
        {
            Id = Guid.NewGuid(),
            UserName = request.Email,
            Email = request.Email,
            PhoneNumber = request.PhoneNumber,
            Address = request.Address,
            FirstName = request.FirstName,
            LastName = request.LastName,
            IsActive = true,
            EmailConfirmed = false,
            AgreedToTermsAt = DateTimeOffset.UtcNow,
            CreatedAt = DateTimeOffset.UtcNow,
        };

        var createResult = await _userManager.CreateAsync(user, request.Password);
        if (!createResult.Succeeded)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                [nameof(request.Password)] = createResult.Errors.Select(e => e.Description).ToArray(),
            });
        }

        var propertyManagerMembership = await ProvisionWorkspaceAsync(
            user.Id, request.WorkspaceName, request.PortfolioSize, cancellationToken);

        await SendActivationEmailAsync(user, cancellationToken);

        return await IssueTokensAsync(user.Id, propertyManagerMembership, clientIp, cancellationToken);
    }

    /// <summary>
    /// US-16: sends a fresh tokenized confirmation link. Same-generic-response,
    /// enumeration-safe pattern as forgot-password.
    /// </summary>
    public async Task ResendActivationAsync(string email, CancellationToken cancellationToken)
    {
        var user = await _userManager.FindByEmailAsync(email);
        if (user is not null && !user.EmailConfirmed)
        {
            await SendActivationEmailAsync(user, cancellationToken);
        }
    }

    /// <summary>US-16: confirms the account from an activation link's UserId + Token.</summary>
    public async Task ActivateAsync(Guid userId, string token, CancellationToken cancellationToken)
    {
        var user = await _userManager.FindByIdAsync(userId.ToString());
        if (user is null)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                [nameof(token)] = ["This activation link is invalid or has expired."],
            });
        }

        var result = await _userManager.ConfirmEmailAsync(user, token);
        if (!result.Succeeded)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                [nameof(token)] = ["This activation link is invalid or has expired."],
            });
        }
    }

    /// <summary>
    /// US-16: same enumeration-safe pattern as resend-activation -- identical outcome
    /// whether or not the email exists, so a caller can't use this to probe valid accounts.
    /// </summary>
    public async Task ForgotPasswordAsync(string email, CancellationToken cancellationToken)
    {
        var user = await _userManager.FindByEmailAsync(email);
        if (user is not null && user.IsActive)
        {
            var token = await _userManager.GeneratePasswordResetTokenAsync(user);
            var link = $"{_frontendBaseUrl}/reset-password?userId={user.Id}&token={Uri.EscapeDataString(token)}";

            await _emailSender.SendAsync(
                user.Email!,
                "Reset your Ten21 password",
                $"<p>Someone requested a password reset for this account. If this was you, " +
                $"click the link below (expires in 24 hours):</p><p><a href=\"{link}\">{link}</a></p>" +
                $"<p>If you didn't request this, you can safely ignore this email.</p>",
                cancellationToken);
        }
    }

    /// <summary>
    /// US-16: resets the password from a reset link's UserId + Token, then revokes every
    /// other active session for the account -- a token issued under the old password
    /// shouldn't silently keep working after a reset.
    /// </summary>
    public async Task ResetPasswordAsync(
        Guid userId, string token, string newPassword, string? clientIp, CancellationToken cancellationToken)
    {
        var user = await _userManager.FindByIdAsync(userId.ToString());
        if (user is null)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                ["Token"] = ["This password reset link is invalid or has expired."],
            });
        }

        var result = await _userManager.ResetPasswordAsync(user, token, newPassword);
        if (!result.Succeeded)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                ["NewPassword"] = result.Errors.Select(e => e.Description).ToArray(),
            });
        }

        await _refreshTokenService.RevokeAllForUserAsync(user.Id, clientIp, cancellationToken);
    }

    /// <summary>US-14/US-16: builds and sends the tokenized activation link. Shared by
    /// Register (first send) and ResendActivation (on-demand).</summary>
    private async Task SendActivationEmailAsync(ApplicationUser user, CancellationToken cancellationToken)
    {
        var token = await _userManager.GenerateEmailConfirmationTokenAsync(user);
        var link = $"{_frontendBaseUrl}/activate?userId={user.Id}&token={Uri.EscapeDataString(token)}";

        await _emailSender.SendAsync(
            user.Email!,
            "Confirm your Ten21 account",
            $"<p>Welcome to Ten21! Confirm your email address by clicking the link below " +
            $"(expires in 24 hours):</p><p><a href=\"{link}\">{link}</a></p>",
            cancellationToken);
    }

    /// <summary>
    /// US-15: Google Sign-In. Verifies the Google ID token server-side, then either
    /// auto-links to an existing account (by prior Google login, falling back to a
    /// verified-email match) or creates a brand-new, passwordless ApplicationUser. A user
    /// with no workspace yet gets an interim token instead of a full AuthResponse.
    /// </summary>
    public async Task<GoogleLoginResult> GoogleLoginAsync(
        string idToken, string? clientIp, CancellationToken cancellationToken)
    {
        var identity = await _googleIdTokenVerifier.VerifyAsync(idToken, cancellationToken);
        if (identity is null || !identity.EmailVerified)
        {
            throw new UnauthorizedException("Invalid Google credential.");
        }

        var user = await _userManager.FindByLoginAsync(GoogleLoginProvider, identity.Subject);

        if (user is null)
        {
            // No prior Google login recorded -- but a verified Google email matching an
            // existing account's email is enough to auto-link. Google already vouches for
            // email ownership (EmailVerified above), so this isn't trusting client input the
            // way it would be for an unverified claim.
            user = await _userManager.FindByEmailAsync(identity.Email);
            if (user is not null)
            {
                var linkResult = await _userManager.AddLoginAsync(
                    user, new UserLoginInfo(GoogleLoginProvider, identity.Subject, GoogleLoginProvider));
                if (!linkResult.Succeeded)
                {
                    throw new ValidationException(new Dictionary<string, string[]>
                    {
                        ["Google"] = linkResult.Errors.Select(e => e.Description).ToArray(),
                    });
                }
            }
        }

        if (user is null)
        {
            user = new ApplicationUser
            {
                Id = Guid.NewGuid(),
                UserName = identity.Email,
                Email = identity.Email,
                EmailConfirmed = true, // Google already verified it
                FirstName = string.IsNullOrWhiteSpace(identity.GivenName) ? "New" : identity.GivenName,
                LastName = string.IsNullOrWhiteSpace(identity.FamilyName) ? "User" : identity.FamilyName,
                IsActive = true,
                CreatedAt = DateTimeOffset.UtcNow,
            };

            // No password -- CheckPasswordAsync will simply always fail for this account,
            // which is correct: Google Sign-In is the only way in for a Google-only user.
            var createResult = await _userManager.CreateAsync(user);
            if (!createResult.Succeeded)
            {
                throw new ValidationException(new Dictionary<string, string[]>
                {
                    ["Email"] = createResult.Errors.Select(e => e.Description).ToArray(),
                });
            }

            await _userManager.AddLoginAsync(
                user, new UserLoginInfo(GoogleLoginProvider, identity.Subject, GoogleLoginProvider));
        }

        var hasWorkspace = await _dbContext.TenantMemberships
            .IgnoreQueryFilters()
            .AnyAsync(tm => tm.UserId == user.Id, cancellationToken);

        if (!hasWorkspace)
        {
            var interimToken = _jwtTokenService.GenerateInterimAccessToken(user.Id, TokenPurposes.ProfileIncomplete);
            return GoogleLoginResult.ForProfileCompletion(
                new ProfileCompletionRequiredResponse(true, interimToken.Value, interimToken.ExpiresAtUtc));
        }

        var membership = await ResolvePrimaryMembershipAsync(user.Id, cancellationToken)
            ?? throw new UnauthorizedException("This account has no property/tenant access configured.");

        var tokens = await IssueTokensAsync(user.Id, membership, clientIp, cancellationToken);
        return GoogleLoginResult.ForTokens(tokens);
    }

    /// <summary>US-15: the second half of a first-time Google signup. The controller has
    /// already checked the caller presented a profile_incomplete interim token before
    /// calling this.</summary>
    public async Task<AuthTokenResult> CompleteProfileAsync(
        Guid userId, CompleteProfileRequest request, string? clientIp, CancellationToken cancellationToken)
    {
        var alreadyHasWorkspace = await _dbContext.TenantMemberships
            .IgnoreQueryFilters()
            .AnyAsync(tm => tm.UserId == userId, cancellationToken);
        if (alreadyHasWorkspace)
        {
            throw new ConflictException("This account already has a workspace.");
        }

        var user = await _userManager.FindByIdAsync(userId.ToString())
            ?? throw new UnauthorizedException("Account no longer exists.");

        user.PhoneNumber = request.PhoneNumber;
        user.Address = request.Address;
        await _userManager.UpdateAsync(user);

        var propertyManagerMembership = await ProvisionWorkspaceAsync(
            userId, request.WorkspaceName, request.PortfolioSize, cancellationToken);

        return await IssueTokensAsync(userId, propertyManagerMembership, clientIp, cancellationToken);
    }

    public async Task<LoginResult> LoginAsync(
        LoginRequest request, string? clientIp, CancellationToken cancellationToken)
    {
        var user = await _userManager.FindByEmailAsync(request.Email);

        // Same generic message regardless of WHICH check fails (no such user, wrong
        // password, inactive account) -- distinguishing them would let a caller enumerate
        // valid email addresses one login attempt at a time.
        if (user is null || !user.IsActive)
        {
            throw new UnauthorizedException(GenericLoginFailureMessage);
        }

        if (await _userManager.IsLockedOutAsync(user))
        {
            throw new UnauthorizedException(
                "Account is temporarily locked due to repeated failed login attempts.");
        }

        if (!await _userManager.CheckPasswordAsync(user, request.Password))
        {
            await _userManager.AccessFailedAsync(user); // counts toward the 5-attempt lockout
            throw new UnauthorizedException(GenericLoginFailureMessage);
        }

        await _userManager.ResetAccessFailedCountAsync(user);

        // US-24: checked BEFORE membership/2FA -- an account provisioned with a temporary
        // password must change it before anything else happens, including a 2FA challenge.
        if (user.MustChangePassword)
        {
            var passwordChallenge = _jwtTokenService.GenerateInterimAccessToken(user.Id, TokenPurposes.PasswordChangePending);
            return LoginResult.ForPasswordChange(
                new PasswordChangeRequiredResponse(true, passwordChallenge.Value, passwordChallenge.ExpiresAtUtc));
        }

        return await CompleteLoginAsync(user, clientIp, cancellationToken);
    }

    /// <summary>
    /// The shared tail of both a normal Login (once MustChangePassword is confirmed false)
    /// and ChangeTempPassword's successful completion: resolve tenant membership, gate on
    /// mandatory 2FA if this role requires it, otherwise issue a real session.
    /// </summary>
    private async Task<LoginResult> CompleteLoginAsync(
        ApplicationUser user, string? clientIp, CancellationToken cancellationToken)
    {
        var membership = await ResolvePrimaryMembershipAsync(user.Id, cancellationToken);
        if (membership is null)
        {
            throw new UnauthorizedException("This account has no property/tenant access configured.");
        }

        // US-17 (email-only, per Founder decision): password alone isn't enough for
        // SuperAdmin/PropertyManager/BoardMember (SECURITY.md §1's mandatory-MFA roles).
        var role = await _roleManager.FindByIdAsync(membership.RoleId.ToString())
            ?? throw new InvalidOperationException(
                $"Role {membership.RoleId} referenced by a TenantMembership no longer exists.");

        if (MandatoryTwoFactorRoles.Values.Contains(role.Name))
        {
            // A fresh, cryptographically random code every call -- no dependency on
            // Identity's built-in Email token provider, whose ~3-minute TOTP step is
            // hardcoded and returns the SAME code for repeated calls within that step.
            var code = RandomNumberGenerator.GetInt32(0, 1_000_000).ToString("D6");
            var codeHash = Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(code)));
            var codeExpiresAtUtc = DateTimeOffset.UtcNow.AddMinutes(5);

            await _emailSender.SendAsync(
                user.Email!,
                "Your Ten21 sign-in code",
                $"<p>Your sign-in code is: <strong>{code}</strong></p>" +
                $"<p>Enter it to finish signing in. This code expires in 5 minutes.</p>",
                cancellationToken);

            var challenge = _jwtTokenService.GenerateTwoFactorChallengeToken(user.Id, codeHash, codeExpiresAtUtc);
            return LoginResult.ForTwoFactor(new TwoFactorRequiredResponse(true, challenge.Value, challenge.ExpiresAtUtc));
        }

        var tokens = await IssueTokensAsync(user.Id, membership, clientIp, cancellationToken);
        return LoginResult.ForTokens(tokens);
    }

    /// <summary>
    /// US-24: the second half of a MustChangePassword-gated login. The controller has
    /// already checked the caller presented a PasswordChangePending challenge token. No
    /// CurrentPassword field -- the challenge token itself already proves knowledge of the
    /// current (temporary) password. Uses RemovePasswordAsync + AddPasswordAsync rather than
    /// ChangePasswordAsync for exactly that reason -- there's nothing left to re-verify.
    /// </summary>
    public async Task<LoginResult> ChangeTempPasswordAsync(
        Guid userId, string newPassword, string? clientIp, CancellationToken cancellationToken)
    {
        var user = await GetUserAsync(userId);

        var removeResult = await _userManager.RemovePasswordAsync(user);
        if (!removeResult.Succeeded)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                ["NewPassword"] = removeResult.Errors.Select(e => e.Description).ToArray(),
            });
        }

        var addResult = await _userManager.AddPasswordAsync(user, newPassword);
        if (!addResult.Succeeded)
        {
            throw new ValidationException(new Dictionary<string, string[]>
            {
                ["NewPassword"] = addResult.Errors.Select(e => e.Description).ToArray(),
            });
        }

        user.MustChangePassword = false;
        await _userManager.UpdateAsync(user);

        return await CompleteLoginAsync(user, clientIp, cancellationToken);
    }

    /// <summary>US-17: the second half of a 2FA-gated login. The controller has already
    /// validated the submitted code against the challenge token's code_hash/code_exp claims
    /// before calling this.</summary>
    public async Task<AuthTokenResult> VerifyTwoFactorAsync(
        Guid userId, string? clientIp, CancellationToken cancellationToken)
    {
        var user = await GetUserAsync(userId);

        var membership = await ResolvePrimaryMembershipAsync(user.Id, cancellationToken)
            ?? throw new UnauthorizedException("This account has no property/tenant access configured.");

        return await IssueTokensAsync(user.Id, membership, clientIp, cancellationToken);
    }

    public async Task<RefreshTokenOutcome> RefreshTokenAsync(
        string rawToken, string? clientIp, CancellationToken cancellationToken)
    {
        RefreshTokenRotationResult rotation;
        try
        {
            rotation = await _refreshTokenService.ValidateAndRotateAsync(rawToken, clientIp, cancellationToken);
        }
        catch (RefreshTokenException ex)
        {
            return RefreshTokenOutcome.Failed(ex.Message);
        }

        // IgnoreQueryFilters is required and deliberate: no ITenantContext is resolved for
        // this request (there's no valid access token yet -- that's the entire point of
        // refresh), so the fail-closed filter from US-01 would otherwise return zero rows.
        //
        // A list, not SingleOrDefaultAsync: SECURITY.docx explicitly supports a user holding
        // more than one role in the SAME tenant. Prefer whichever membership is IsPrimary,
        // same tie-break Login already uses, so a refreshed token keeps the same role the
        // original login issued.
        var memberships = await _dbContext.TenantMemberships
            .IgnoreQueryFilters()
            .Where(tm => tm.UserId == rotation.UserId && tm.TenantId == rotation.TenantId)
            .ToListAsync(cancellationToken);

        var membership = memberships.FirstOrDefault(m => m.IsPrimary) ?? memberships.FirstOrDefault();

        if (membership is null)
        {
            // Membership was revoked between issuance and this refresh (e.g. an admin
            // removed this person from the property). The old refresh token was already
            // consumed by ValidateAndRotateAsync above; deny the new access token too.
            return RefreshTokenOutcome.Failed("Tenant access for this account has changed. Please log in again.");
        }

        var response = await BuildAuthResponseAsync(rotation.UserId, membership, cancellationToken);
        return RefreshTokenOutcome.Succeeded(new AuthTokenResult(response, rotation.NewRawToken));
    }

    public Task RevokeAsync(string rawToken, string? clientIp, CancellationToken cancellationToken) =>
        _refreshTokenService.RevokeAsync(rawToken, clientIp, cancellationToken);

    /// <summary>
    /// Shared by Register and CompleteProfile: creates the new Tenant (workspace) and a
    /// root TenantMembership pair -- PropertyManager (IsPrimary = true) and PropertyOwner
    /// (IsPrimary = false) -- for userId. See Register's doc comment for why a self-service
    /// landlord gets both roles.
    /// </summary>
    private async Task<TenantMembership> ProvisionWorkspaceAsync(
        Guid userId, string workspaceName, int portfolioSize, CancellationToken cancellationToken)
    {
        var tenant = new Tenant
        {
            Id = Guid.NewGuid(),
            Name = workspaceName,
            PortfolioSize = portfolioSize,
            OrganizationId = null,
            CreatedAt = DateTimeOffset.UtcNow,
        };
        // Tenant is not ITenantScopedEntity (it sits above the tenant boundary) -- no active
        // tenant context is needed for its own insert.
        _dbContext.Tenants.Add(tenant);
        await _dbContext.SaveChangesAsync(cancellationToken);

        var propertyManagerRole = await _roleManager.FindByNameAsync(RoleNames.PropertyManager)
            ?? throw new InvalidOperationException("PropertyManager role not seeded -- has RoleSeeder run?");
        var propertyOwnerRole = await _roleManager.FindByNameAsync(RoleNames.PropertyOwner)
            ?? throw new InvalidOperationException("PropertyOwner role not seeded -- has RoleSeeder run?");

        // TenantMembership IS ITenantScopedEntity -- the fail-closed insert guard (US-01)
        // requires an active tenant context before either row below can be added.
        _tenantContext.SetTenant(tenant.Id);

        var propertyManagerMembership = new TenantMembership
        {
            Id = Guid.NewGuid(),
            TenantId = tenant.Id,
            UserId = userId,
            RoleId = propertyManagerRole.Id,
            IsPrimary = true,
            CreatedAt = DateTimeOffset.UtcNow,
        };
        _dbContext.TenantMemberships.Add(propertyManagerMembership);
        _dbContext.TenantMemberships.Add(new TenantMembership
        {
            Id = Guid.NewGuid(),
            TenantId = tenant.Id,
            UserId = userId,
            RoleId = propertyOwnerRole.Id,
            IsPrimary = false,
            CreatedAt = DateTimeOffset.UtcNow,
        });
        await _dbContext.SaveChangesAsync(cancellationToken);

        return propertyManagerMembership;
    }

    private async Task<TenantMembership?> ResolvePrimaryMembershipAsync(Guid userId, CancellationToken cancellationToken)
    {
        var memberships = await _dbContext.TenantMemberships
            .IgnoreQueryFilters() // same bootstrap reasoning as the refresh-token lookup above
            .Where(tm => tm.UserId == userId)
            .ToListAsync(cancellationToken);

        return memberships.FirstOrDefault(m => m.IsPrimary) ?? memberships.FirstOrDefault();
    }

    private async Task<AuthTokenResult> IssueTokensAsync(
        Guid userId, TenantMembership membership, string? clientIp, CancellationToken cancellationToken)
    {
        var rawRefreshToken = await _refreshTokenService.IssueAsync(
            userId, membership.TenantId, clientIp, cancellationToken);

        var response = await BuildAuthResponseAsync(userId, membership, cancellationToken);
        return new AuthTokenResult(response, rawRefreshToken);
    }

    private async Task<AuthResponse> BuildAuthResponseAsync(
        Guid userId, TenantMembership membership, CancellationToken cancellationToken)
    {
        var role = await _roleManager.FindByIdAsync(membership.RoleId.ToString())
            ?? throw new InvalidOperationException($"Role {membership.RoleId} referenced by a TenantMembership no longer exists.");

        var tenant = await _dbContext.Tenants
            .IgnoreQueryFilters() // Tenant isn't ITenantScopedEntity, but be explicit rather than rely on that
            .SingleAsync(t => t.Id == membership.TenantId, cancellationToken);

        var accessToken = _jwtTokenService.GenerateAccessToken(
            userId, membership.TenantId, tenant.OrganizationId, role.Name!);

        return new AuthResponse(
            accessToken.Value, accessToken.ExpiresAtUtc, membership.TenantId, tenant.OrganizationId, role.Name!);
    }

    /// <summary>US-15/US-17/US-24 shared helper: resolves an ApplicationUser by id for a
    /// caller the controller has already authenticated via an interim token's user_id
    /// claim.</summary>
    private async Task<ApplicationUser> GetUserAsync(Guid userId) =>
        await _userManager.FindByIdAsync(userId.ToString())
            ?? throw new UnauthorizedException("Account no longer exists.");
}
